IndexShardIT#testAutomaticRefresh test failure

Test failed on master, doesn't reproduce for me.

REPRODUCE WITH: ./gradlew :server:integTest --tests "org.elasticsearch.index.shard.IndexShardIT.testAutomaticRefresh" \
  -Dtests.seed=FBE5E0B07A2F06E7 \
  -Dtests.security.manager=true \
  -Dtests.locale=nn-NO \
  -Dtests.timezone=Asia/Colombo \
  -Dcompiler.java=12 \
  -Druntime.java=11

Log: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=debian-9&&immutable/203/console
Build Scans: https://gradle-enterprise.elastic.co/s/tg3ytepmbfprw


java.lang.AssertionError: Count is 0 hits but 1 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
at org.junit.Assert.fail(Assert.java:88)
at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:254)
at org.elasticsearch.index.shard.IndexShardIT.testAutomaticRefresh(IndexShardIT.java:727)

--- a/server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java
+++ b/server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java
@@ -709,10 +709,16 @@ public class IndexShardIT extends ESSingleNodeTestCase {
             // with ZERO we are guaranteed to see the doc since we will wait for a refresh in the background
             assertFalse(hasRefreshed);
             assertTrue(shard.isSearchIdle());
-        } else if (randomTimeValue == null){
-            // with null we are guaranteed to see the doc since do execute the refresh.
-            // we can't assert on hasRefreshed since it might have been refreshed in the background on the shard concurrently
-            assertFalse(shard.isSearchIdle());
+        } else {
+            if (randomTimeValue == null) {
+                assertFalse(shard.isSearchIdle());
+            }
+            // we can't assert on hasRefreshed since it might have been refreshed in the background on the shard concurrently.
+            // and if the background refresh wins the refresh race (both call maybeRefresh), the document might not be visible
+            // until the background refresh is done.
+            if (hasRefreshed == false) {
+                ensureNoPendingScheduledRefresh(indexService.getThreadPool());
+            }
         }
         CountDownLatch started = new CountDownLatch(1);
         Thread t = new Thread(() -> {

